Adds extra elements to RSS items.

philnash opened the pull request:
## Changes

I found I wanted to add categories to my RSS feed, but it wasn't supported, aside from through custom data. I thought it would be better for the RSS generator to have types for the available elements than creating my own XML by hand.

I also noticed that the `<guid>` element is always set to the item's permalink, so we can set `<guid isPermaLink="true">` too.

## Testing

Unit tests with an RSS feed with an item using all available options

## Docs

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
/cc @withastro/maintainers-docs for feedback!

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

## More in depth description 

The RSS generation module only currently generates a subset of the available fields in the [RSS spec](https://validator.w3.org/feed/docs/rss2.html#hrelementsOfLtitemgt) opting to allow users to generate their own XML and supply it via the `customData` property.

Since the spec is not very long, I thought it would help to expand the fields available for the `<item>` to the full set. This will allow users to add categories (tags) easily, which is what I was missing, plus links to comments, the authors email address, links to a source feed and media objects (great for podcasts).


The fields added are:

```
/** Categories or tags related to the item */
categories?: string[];

/** The item author's email address */
author?: string;

/** A URL of a page for comments relating to the item */
comments?: string;

/** The RSS channel that the item came from */
source?: { url: string, title: string }

/** A media object that belongs to the item */
enclosure?: { url: string, length: number, type: string }
 ```

This completes the spec for the items. There are additional fields that can be added to the channel as a whole, but I wanted to make small changes, especially as a first proposal. I wasn’t fully aware of the RFC process, and will look into it for future contributions.

philnash commented:
The Windows/Node 16 seems to have just failed due to timeout. Can someone rerun them for me? I can't restart the job myself.

philnash commented:
Apologies for keeping adding things to this PR. I actually believe it is complete now (after I updated the types and the README with the last commit).

Would love a review or to know what you think? Thanks!

ematipico commented on packages/astro-rss/src/index.ts:
```diff
@@ -47,6 +47,16 @@ type RSSFeedItem = {
+	/** The item author's email address */
+	author?: z.infer<typeof rssSchema>['author'];
+	/** A URL of a page for comments related to the item */
+	comments?: z.infer<typeof rssSchema>['comments'];
```
nit: `commentsUrl`?

philnash replied:
That's a good idea. I was trying to stay close to the RSS field names. Though I did change `category` to `categories` to deal with the plural. This would make it more obvious and I'll update that.

ematipico commented on packages/astro-rss/README.md:
```diff
@@ -128,6 +128,16 @@ type RSSFeedItem = {
+  /** The RSS channel that the item came from */
+  source?: { url: string, title: string }
+  /** A media object that belongs to the item */
+  enclosure?: { url: string, length: number, type: string }
```
We should explain/document all these fields. It's not clear what `length` and `type` are

ematipico commented on packages/astro-rss/README.md:
```diff
@@ -128,6 +128,16 @@ type RSSFeedItem = {
+  /** A URL of a page for comments relating to the item */
+  comments?: string;
+  /** The RSS channel that the item came from */
+  source?: { url: string, title: string }
```
We should explain the fields. It's not clear, for example, what `title` is

ematipico commented on packages/astro-rss/README.md:
```diff
@@ -128,6 +128,16 @@ type RSSFeedItem = {
   /** Append some other XML-valid data to this item */
   customData?: string;
+  /** Categories or tags related to the item */
+  categories?: string[];
```
Maybe we could provide some examples of these categories?

philnash commented:
I have updated the docs in the README, upgrading the description of an `RSSFeedItem` from a quick type, to full descriptions of the properties. And added an example object.

I also updated `comments` to `commentsUrl`.

There is a failing test that appears to be unrelated (it's to do with JSX). The Windows and Ubuntu versions of those tests passed, so I don't know why the Mac version is flaking on me.

ElianCodes commented on packages/astro-rss/README.md:
````diff
@@ -202,6 +185,112 @@ export const get = () => rss({
 });
 ```
 
+## `RSSFeedItem`
````
I really like how well you've restructured the docs and how well you explained this

ElianCodes reviewed (commented):
I really like this! The only thing I'm missing is explaining some differences in between the new and old version! Maybe you should add a `upgrading to vX.x.x` section?

sarah11918 commented:
Heads up that I'm in the middle of a review on this one!

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `object (optional)`
+
+Items that are republished from other publications may define a `source` which defines the `title` and `url` of the original feed in which it was published.
```
``````suggestion
An object that defines the `title` and `url` of the original feed for items that have been republished from another source. Both are required propeties of `source` for proper attribution. 

```js
const item = {
	title: "Alpha Centauri: so close you can touch it",
  link: "/blog/alpha-centuari",
  pubDate: new Date("2023-06-04"),
	source:
		{
			title: "The Galactic Times",
			url: "https://galactictimes.space/feed.xml"
		}
}
```
``````
I also think an example (choose one that makes sense, formatted properly!) would make sense here.

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (required)`
+
+If you define a `source` you must define that source's `title`. It is the name of the original feed in which the item was published.
```
```suggestion
The name of the original feed in which the item was published. (Note that this is the the feed's title, not the individual article title.)
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Items that are republished from other publications may define a `source` which defines the `title` and `url` of the original feed in which it was published.
+
+#### `title`
```
```suggestion
#### `source.title`
```
I'm wondering if this is helpful, especially as a heading that might show up in Algolia if someone searches for `title` in another context, like frontmatter of a Markdown file. Also helps to connect the properties to their object.

philnash replied:
Good idea, thanks.

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (required)`
+
+If you define a `source` you must also define that source's `url` which identifies the URL of the original feed in which the item was published.
```
```suggestion
The URL of the original feed in which the item was published.
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+If you define a `source` you must define that source's `title`. It is the name of the original feed in which the item was published.
+
+#### `url`
```
```suggestion
#### `source.url`
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `object (optional)`
+
+Items that include media as part of the feed, like a podcast, can define an `enclosure` which is made of three required fields, a `url`, `length`, and `type`.
```
``````suggestion
An object to specify properties for an included media source (e.g. a podcast) with three required values: `url`, `length`, and `type`.

```js
const item = {
	title: "Alpha Centauri: so close you can touch it",
  link: "/blog/alpha-centuari",
  pubDate: new Date("2023-06-04"),
	enclosure:
		{
			url: "https://mydomain/media/alpha-centauri.aac",
			length: 124568,
			type: "audio/aac"
		}
}
```
``````
I also think an example would be helpful here! Note, if the URL is from your own site, do you still need to use a full URL here? This is just an example. Adjust to make this correct!

philnash replied:
The RSS spec says that you do need this to be a full URL. However, we can treat this like `link`, check if it's a valid URL and if it isn't run it through `createCanonicalURL` to make a full URL for a relative path.

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Items that include media as part of the feed, like a podcast, can define an `enclosure` which is made of three required fields, a `url`, `length`, and `type`.
+
+#### `url`
```
```suggestion
#### `enclosure.url`
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (required)`
+
+The `url` field for the `enclosure` defines a URL where the media can be found.
```
```suggestion
The URL where the media can be found.
```
If you only need to use e.g. `/media/filename` for media hosted on your own site, and not the full URL, then I would also add a sentence mentioning that full URLs are only required for external sources. (And, make sure the example code in `enclosure` demonstrates a correct URL!)

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+The `url` field for the `enclosure` defines a URL where the media can be found.
+
+#### `length`
```
```suggestion
#### `enclosure.length`
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `number (required)`
+
+The `length` field defines the size of the file found at the `url` in bytes.
```
```suggestion
The size of the file found at the `url` in bytes.
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+The `length` field defines the size of the file found at the `url` in bytes.
+
+#### `type`
```
```suggestion
#### `enclosure.type`
```

sarah11918 commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (required)`
+
+The `type` field defines the MIME type for the media item found at the `url`.
```
```suggestion
The [MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) for the media item found at the `url`.
```

sarah11918 reviewed (commented):
Thanks so much for this @philnash! I love me some RSS and I actually can't wait to set this up on my own blog! :raised_hands: 

It was also so easy to edit something so thoughtfully and carefully written. Much appreciated! You'll see a lot of "suggestions" but it's mostly just defining all the properties in a uniform way, starting with a literal description of what the thing is. (Instead of, for example, "If you want to...")

We prefer to have reference definitions start with "Just the facts, ma'am"... and then we often find that we don't *need* a lot of the introductory words. So, it becomes easier for the reader to scan and take in the key information more quickly. I've updated your definitions in that style, to give you an idea. But, sometimes a nuance is lost, so don't just take my changes at face value, and please DO correct anything that isn't quite right! These are just examples of the format I'd use in our docs.

I also indicated a couple of places where I felt a code example would be helpful. Again, please edit those for correctness, as I really just wanted to illustrate what I'd hope to see there.

Ping me directly with any questions, or to review your changes. And again, thank you for such an amazing contribution, and documentation that only needed some structural work to be consistent with our docs!  :partying_face:

philnash commented:
Thanks so much for the reviews on this. I have been traveling for speaking at a conference, so haven't had the time to go over it all. But I should be be able to go through it all this week.

philnash commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (optional)`
+
+If the item is complete within itself, that is you are publishing the full content of the item in the feed, the `description` field may contain the full text (entity-encoded HTML is permitted). If the item is a stub, then the `description` may contain a synopsis of the item.
```
Yeah, RSS is old and a mess! And feed readers have inconsistent implementations. Such fun!

philnash commented on packages/astro-rss/README.md:
```diff
@@ -202,6 +185,112 @@ export const get = () => rss({
+
+Type: `string (optional)`
+
+Useful for multi-author blogs, the `author` field provides the email address of the person who wrote the item.
```
Agreed, but that's what the spec says. 🤷‍♂️

philnash commented:
Ok, updated the docs and allowed for URLs to be relative (and turned into a full URL using `createCanonicalURL`.

Let me know if this is looking good now, thanks!

philnash commented:
@ElianCodes 

> I really like this! The only thing I'm missing is explaining some differences in between the new and old version! Maybe you should add a `upgrading to vX.x.x` section?

This change only adds optional fields, so there's nothing you have to do to upgrade. Do you think it needs an upgrading section?

Yan-Thomas reviewed (approved):
Docs LGTM!

philnash commented:
Hey team, is there anything else I need to do here? Or does this just need to wait for a good time to merge?

Is the build fail an issue? It seemed flaky to me, but it does put a big red cross against the PR. Can someone rerun the test to see if it will pass?

Thanks

Yan-Thomas commented:
> Hey team, is there anything else I need to do here? Or does this just need to wait for a good time to merge?

Hey @philnash, a good part of the team is coming back and recovering from a trip last week, they will reach this PR soon and properly review the feature, so you don't need to do anything for now. And yes, the test looks flaky, I'll rerun but it's possible it might continue failing, but no need to worry, it's going to be properly analyzed by the team.

philnash commented:
Thanks @Yan-Thomas, didn't realise the team had been away. Hope it was a good trip. And yay! The tests passed!

ematipico reviewed (changes requested):
There is one thing missing that I didn't see before. You have to create a changest via `pnpm changeset`. I think this is a `minor` change because it adds new features and it doesn't break the existing behaviour.

ematipico reviewed (approved):
Thank you! Looks good to me!
